warn/types.go

```diff
@@ -0,0 +1,146 @@
+package warn
+
+import (
+	"github.com/bazelbuild/buildtools/build"
+	"github.com/bazelbuild/buildtools/bzlenv"
+)
+
+// Type describes an expression type in Starlark.
+type Type int
+
+// List of known types
+const (
+	Unknown Type = iota
+	Bool
+	Depset
+	Dict
+	Int
+	None
+	String
+)
+
+func (t Type) String() string {
+	return [...]string{
+		"unknown",
+		"bool",
+		"depset",
+		"dict",
+		"int",
+		"none",
+		"string",
+	}[t]
+}
+
+func detectTypes(f *build.File) map[build.Expr]Type {
+	variables := make(map[int]Type)
+	result := make(map[build.Expr]Type)
+
+	var walk func(e *build.Expr, env *bzlenv.Environment)
+	walk = func(e *build.Expr, env *bzlenv.Environment) {
+		// Postorder: determining types of subnodes may help with this node's type
+		walkOnce(*e, env, walk)
+
+		nodeType := Unknown
+
+		switch node := (*e).(type) {
+		case *build.StringExpr:
+			nodeType = String
+		case *build.DictExpr:
+			nodeType = Dict
+		case *build.LiteralExpr:
+			nodeType = Int
+		case *build.Comprehension:
+			if node.Curly {
+				nodeType = Dict
+			}
+		case *build.CallExpr:
+			if ident, ok := (node.X).(*build.Ident); ok {
+				switch ident.Name {
+				case "depset":
+					nodeType = Depset
+				case "dict":
+					nodeType = Dict
+				}
+			}
+		case *build.ParenExpr:
+			nodeType = result[node.X]
+		case *build.Ident:
+			switch node.Name {
+			case "True", "False":
+				nodeType = Bool
+				break
+			case "None":
+				nodeType = None
+				break
+			}
+			binding := env.Get(node.Name)
+			if binding != nil {
+				if t, ok := variables[binding.ID]; ok {
+					nodeType = t
+				}
+			}
+		case *build.BinaryExpr:
+			switch node.Op {
+			case "=", "+=", "-=", "*=", "/=", "//=", "%=", "|=":
+				// Assignments
+				t, ok := result[node.Y]
+				if !ok {
+					break
+				}
+				ident, ok := (node.X).(*build.Ident)
+				if !ok {
+					break
+				}
+				binding := env.Get(ident.Name)
+				if binding == nil {
+					break
+				}
+				variables[binding.ID] = t
+
+			case ">", ">=", "<", "<=", "==", "!=", "in", "not in":
+				// Boolean
+				nodeType = Bool
+
+			case "+", "-", "*", "/", "//", "%", "|":
+				// We assume these operators can only applied to expressions of the same type and
+				// preserve the type
+				if t, ok := result[node.X]; ok {
+					nodeType = t
+				} else if t, ok := result[node.Y]; ok {
+					nodeType = t
+				}
+			}
+		}
+
+		if nodeType != Unknown {
```

Comment:
opt: make this a defer right after the declaration, then in the switch, use return instead of break

Reply:
Done.